search/includes/classes/class-versioning.php

```diff
@@ -745,4 +759,176 @@ public function action__ep_delete_indexable( $object_id, $indexable_slug ) {
 		// Clear the flag to return to normal
 		$this->is_doing_object_delete = false;
 	}
+
+	private function is_self_heal_ongoing() {
+		return 1 === wp_cache_get( self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_KEY, self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_GROUP );
+	}
+
+	private function mark_self_heal_ongoing() {
+		wp_cache_set( self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_KEY, 1, self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_GROUP, self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_TTL );
+	}
+
+	private function mark_self_heal_finished() {
+		wp_cache_delete( self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_KEY, self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_GROUP );
+	}
+
+	/**
+	 * Check if the versions are persisted correctly. If not recreate them.
+	 */
+	public function maybe_self_heal() {
+		if ( $this->is_self_heal_ongoing() ) {
+			return;
+		}
+		$this->mark_self_heal_ongoing();
+
+		if ( $this->is_versioning_valid() ) {
+			$this->mark_self_heal_finished();
+			return;
+		}
+
+		$indicies = $this->get_all_accesible_indicies();
+		if ( is_wp_error( $indicies ) ) {
+			$this->mark_self_heal_finished();
+			return;
+		}
+		$versioning = $this->reconstruct_versioning_option( $indicies );
+
+		update_option( self::INDEX_VERSIONS_OPTION, $versioning );
+
+		$this->mark_self_heal_finished();
+	}
+
+	public function is_versioning_valid() {
+		$all_versions = get_option( self::INDEX_VERSIONS_OPTION, array() );
+
+		if ( ! is_array( $all_versions ) || count( $all_versions ) == 0 ) {
+			return false;
+		}
+
+		foreach ( $all_versions as $versions ) {
+			if ( ! is_array( $versions ) || count( $versions ) == 0 ) {
+				return false;
+			}
+
+			$active_statuses = wp_list_pluck( $versions, 'active' );
+
+			$array_index = array_search( true, $active_statuses, true );
+
+			if ( false === $array_index ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	public function get_all_accesible_indicies() {
+		$response = $this->elastic_search_instance->remote_request( '_cat/indices?format=json' );
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$response_code = (int) wp_remote_retrieve_response_code( $response );
+
+		if ( $response_code >= 400 ) {
+			return new \WP_Error(
+				'failed-to-fetch-indicies',
+				sprintf( 'Request failed to fetch indicies with status %s', $response_code )
+			);
+		}
+
+		$response_body_json = wp_remote_retrieve_body( $response );
+		$response_body = json_decode( $response_body_json, true );
+		$found_indices = [];
+
+		if ( ! is_array( $response_body ) ) {
+			return $found_indices;
+		}
+
+		foreach ( $response_body as $index_obj ) {
+			if ( is_array( $index_obj ) && isset( $index_obj['index'] ) ) {
+				$found_indices[] = $index_obj['index'];
+			}
+		}
+
+		return $found_indices;
+	}
+
+	public function reconstruct_versioning_option( $indicies ) {
+		if ( ! is_array( $indicies ) ) {
+			return [];
```

Comment:
Will this cause an infinite loop of reconstruction attempts? Since we return an empty array, that will get saved into the option, then `is_versioning_valid()` will think it's invalid (empty) and try to re-construct it.

I think we need to set a longer lock when we fail to properly re-construct the option. So if we fail, we can set the lock for say 5 minutes so we don't keep hammering ES with `_cat/indices` requests on every request. We also need to log that condition to the PHP log and probably the log2logstash log so we can surface the problem.

Reply:
@pschoffer I think this is still a problem - an empty array will cause line 813 to return false, marking it as invalid, so the process will just keep going over and over on every request. In theory this only happens if ES is down or we can't load the indexes for some other reason, but we should be defensive here and have a "circuit breaker" that prevents this from trying to rebuild over and over.

Reply:
I added the longer (10 m) backoff and slack alert, but not sure how to handle that differently.

Reply:
We could return `false` or a `WP_Error` here and check for that value in the calling code - if it gets an error, it sets a longer lock to prevent this from running over and over (and failing). Not necessary for this PR but IMO we should do it before we take this out of dry run mode.